### PR TITLE
Procedure for dropping statistics

### PR DESCRIPTION
**Pull request by wendigo**
This PR adds procedure to drop hive statistics

### CODE REVIEW

**Comment by findepi** [comment]
cc @aalbu

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 1]
```diff
@@ -630,6 +630,18 @@ specify a subset of columns to be analyzed via the optional ``columns`` property
 This query collects statistics for columns ``col_1`` and ``col_2`` for the partition
 with keys ``p2_value1, p2_value2``.
 
+
```
Comment (findepi) [inline presto-docs/src/main/sphinx/connector/hive.rst]: redundant empty line

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 2]
```diff
@@ -630,6 +630,18 @@ specify a subset of columns to be analyzed via the optional ``columns`` property
...
+Note that if statistics were previously collected for all columns, they need to be dropped
+before re-analyzing just a subset::
+
+    CALL system.drop_stats(schema_name, table_name, ARRAY[ARRAY['p2_value1', 'p2_value2']])
+
+Statistics collection is supported for Hive Metastore and Amazon Glue.
```
Comment (findepi) [inline presto-docs/src/main/sphinx/connector/hive.rst]: that's not true, remove

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 3]
```diff
@@ -630,6 +630,18 @@ specify a subset of columns to be analyzed via the optional ``columns`` property
...
+
+Amazon Glue Support
+-------------------
+Configuring and using Presto with AWS Glue is described in the :doc:`/aws/glue` documentation section.
```
Comment (findepi) [inline presto-docs/src/main/sphinx/connector/hive.rst]: Does not exist; remove

File: presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                .map(HiveColumnHandle.class::cast)
+                .filter(HiveColumnHandle::isPartitionKey)
+                .map(HiveColumnHandle::getName)
+                .collect(toImmutableList());
+
+        if (partitionValues != null) {
+            // drop stats for specified partitions
+            List<List<String>> partitionStringValues = partitionValues.stream()
+                    .map(DropStatsProcedure::validateParameterType)
+                    .collect(ImmutableList.toImmutableList());
```
Comment (ebyhr) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java]:
```suggestion
                    .collect(toImmutableList());
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                                table,
+                                partitionName,
+                                stats -> PartitionStatistics.empty())));
+            }
+        }
+
+        hiveMetadata.commit();
+    }
+
+    private static List<String> validateParameterType(Object param)
```
Comment (ebyhr) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java]:
```suggestion
    private static List<String> validateParameterType(Object parameter)
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    {
+        if (param == null) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Null partition value");
+        }
+        if (param instanceof List) {
+            List<?> list = (List<?>) param;
```
Comment (ebyhr) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java]: inline?

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 7]
```diff
@@ -722,6 +734,13 @@ Procedures
     * ``DROP``: drop any partitions that exist in the metastore, but not on the file system.
     * ``FULL``: perform both ``ADD`` and ``DROP``.
 
+* ``system.drop_stats(schema_name, table_name, partitions)``
```
Comment (aalbu) [inline presto-docs/src/main/sphinx/connector/hive.rst]: The `partitions` argument should be named `partition_values`. My bad.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "No partitions provided");
+        }
+        if (partitionColumns.isEmpty()) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Cannot specify partition values for an unpartitioned table");
+        }
+        partitionValues.forEach(value -> {
+            if (value.size() != partitionColumns.size()) {
+                throw new PrestoException(
+                        INVALID_PROCEDURE_ARGUMENT,
+                        format("Partition values %s don't not match the number of partition columns (%s)", value, partitionColumns.size()));
```
Comment (ebyhr) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java]: typo: "don't not"

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 9]
```diff
@@ -4575,6 +4575,314 @@ public void testAnalyzeUnpartitionedTableWithColumnSubset()
...
+                "No partitions provided");
+        assertQueryFails(
+                format("CALL system.drop_stats('%s', '%s', ARRAY[ARRAY['p2', '7', 'dummy']])", TPCH_SCHEMA, partitionedTableName),
+                ".*don't not match the number of partition columns.*");
+        assertQueryFails(
+                format("CALL system.drop_stats('%s', '%s', ARRAY[ARRAY['WRONG', 'KEY']])", TPCH_SCHEMA, partitionedTableName),
+                "Partition '.*' not found");
+
+        assertUpdate(format("DROP TABLE %s", unpartitionedTableName));
+        assertUpdate(format("DROP TABLE %s", partitionedTableName));
```
Comment (ebyhr) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: nit: We can use `+` oeprator in this situation. (No need to fix)
```java
assertUpdate("DROP TABLE " + partitionedTableName);
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+                // for non-partitioned tables, just wipe table stats
+                metastore.updateTableStatistics(
+                        new HiveIdentity(session.getIdentity()),
+                        schema,
+                        table,
+                        stats -> PartitionStatistics.empty());
+            }
+            else {
+                // the table is partitioned; remove stats for every partition
+                metastore.getPartitionNames(new HiveIdentity(session.getIdentity()), handle.getSchemaName(), handle.getTableName())
```
Comment (ebyhr) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java]: nit: We can suppress NPE warning by
```suggestion
                metastore.getPartitionNames(new HiveIdentity(session.getIdentity()), schema, table)
```

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 11]
```diff
@@ -768,6 +787,13 @@ Add an empty partition to the ``page_views`` table::
         partition_columns => ARRAY['ds', 'country'],
         partition_values => ARRAY['2016-08-09', 'US']);
 
+Drop stats for a partition of the ``page_views`` table::
+
+    CALL system.drop_stats((
```
Comment (ebyhr) [inline presto-docs/src/main/sphinx/connector/hive.rst]:
```suggestion
    CALL system.drop_stats(
```

**Comment by wendigo** [comment]
Fixed @ebyhr @findepi

File: presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    private static List<String> validateParameterType(Object param)
+    {
+        if (param == null) {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Null partition value");
+        }
+
+        if (param instanceof List) {
+            return ((List<?>) param).stream().map(String.class::cast).collect(Collectors.toList());
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java]:
```suggestion
            return ((List<?>) param).stream()
                .map(String.class::cast)
                .collect(toImmutableList());
```

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 13]
```diff
@@ -3972,7 +3972,7 @@ public void testCollectColumnStatisticsOnCreateTable()
                         "('p_varchar', 0E0, 0E0, 0E0, null, null, null), " +
                         "(null, null, null, null, 0E0, null, null)");
 
-        assertUpdate(format("DROP TABLE %s", tableName));
+        assertUpdate("DROP TABLE " + tableName);
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Unrelated change, please extract to separate commit (within this PR), eg `Improve formatting`

Comment (wendigo) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Done

File: presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java (hunk 12 above)
Comment (wendigo) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/DropStatsProcedure.java]: Done
